velox/functions/sparksql/Decimal.cpp

```diff
@@ -0,0 +1,328 @@
+    auto fromType = args[0]->type();
+    auto toType = args[1]->type();
+    auto result =
+        resultRef->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto decimalValue = decodedArgs.at(0);
+    VELOX_CHECK(decodedArgs.at(2)->isConstantMapping());
+    auto nullOnOverflow = decodedArgs.at(2)->valueAt<bool>(0);
+
+    const auto& fromPrecisionScale = getDecimalPrecisionScale(*fromType);
+    const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+    rows.applyToSelected([&](int row) {
+      auto rescaledValue = DecimalUtil::rescaleWithRoundUp<TInput, TOutput>(
+          decimalValue->valueAt<TInput>(row),
+          fromPrecisionScale.first,
+          fromPrecisionScale.second,
+          toPrecisionScale.first,
+          toPrecisionScale.second,
+          nullOnOverflow);
+      if (rescaledValue.has_value()) {
+        result[row] = rescaledValue.value();
+      } else {
+        resultRef->setNull(row, true);
+      }
+    });
+  }
+};
+
+template <typename TInput>
+class MakeDecimalFunction final : public exec::VectorFunction {
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& resultRef) const final {
+    VELOX_CHECK_EQ(args.size(), 3);
+    auto fromType = args[0]->type();
+    auto toType = args[1]->type();
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto unscaledVec = decodedArgs.at(0);
+    VELOX_CHECK(decodedArgs.at(1)->isConstantMapping());
+    VELOX_CHECK(decodedArgs.at(2)->isConstantMapping());
+    auto nullOnOverflow = decodedArgs.at(2)->valueAt<bool>(0);
+    const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+    auto precision = toPrecisionScale.first;
+    auto scale = toPrecisionScale.second;
+    if (precision <= 18) {
+      context.ensureWritable(
+          rows,
+          SHORT_DECIMAL(
+              static_cast<uint8_t>(precision), static_cast<uint8_t>(scale)),
+          resultRef);
+      auto result = resultRef->asUnchecked<FlatVector<UnscaledShortDecimal>>()
+                        ->mutableRawValues();
+      rows.applyToSelected([&](int row) {
+        auto unscaled = unscaledVec->valueAt<int64_t>(row);
+
+        if (UnscaledShortDecimal::valueInRange(unscaled)) {
+          result[row] = UnscaledShortDecimal(unscaled);
+        } else {
+          if (nullOnOverflow) {
+            resultRef->setNull(row, true);
+          } else {
+            VELOX_USER_FAIL("Unscaled value overflow for precision");
+          }
+        }
+      });
+
+    } else {
+      context.ensureWritable(
+          rows,
+          LONG_DECIMAL(
+              static_cast<uint8_t>(precision), static_cast<uint8_t>(scale)),
+          resultRef);
+      auto result = resultRef->asUnchecked<FlatVector<UnscaledShortDecimal>>()
```

Comment:
@jinchengchenghh Just saw this in case. Is `UnscaledShortDecimal` supposed to be `UnscaledLongDecimal`?